pkg/3scale/amp/operator/system_reconciler.go

```diff
@@ -114,16 +114,19 @@ func NewSystemReconciler(baseAPIManagerLogicReconciler BaseAPIManagerLogicReconc
 }
 
 func (r *SystemReconciler) reconcileFileStorage(system *component.System) error {
-	if r.apiManager.Spec.System != nil && r.apiManager.Spec.System.FileStorageSpec != nil {
-		if r.apiManager.Spec.System.FileStorageSpec.PVC != nil {
-			return r.reconcileSharedStorage(system.SharedStorage())
-		} else if r.apiManager.Spec.System.FileStorageSpec.S3 != nil {
-			return r.reconcileS3AWSSecret(system.S3AWSSecret())
-		} else {
-			return fmt.Errorf("No FileStorage spec specified. FileStorage is mandatory")
-		}
+	if r.apiManager.Spec.System.FileStorageSpec.PVC != nil && r.apiManager.Spec.System.FileStorageSpec.S3 != nil {
```

Comment:
I still think we should really remove validation from the reconcilers if possible. As I see reconcilers should already receive consistent state of the CR and ensuring that should not be responsability of the reconcilers as I commented in more detail in https://github.com/3scale/3scale-operator/pull/275/files#r347341179

Reply:
Definitely doing validations in a function meant to set defaults does not fit in any best patterns guide

Validations should be done in a single place at the beginning, or by each component reconciler, like it is being doing already.

It is ok for you if I remove validation fix and leave it 'ugly' but still doing other items?
* Fix storageClassName property location
* CR defaults cannot be mysql. When external databases are configured, this leads to inconsistent information. Changed defaults set for the CR. Mysql is still the default database if none configured
* ~~CR validations taken to reconciliation logic and removed from setting defaults logic~~

Reply:
* Ok to fix storageClassName property location :+1: .
* Ok to remove validation "fix" and leave it "ugly". I'd prefer not mixing reconcile logic and mixing defaults logic. As I commented to me conceptually a reconciler should start having an already validated/consistent CR

About "CR defaults cannot be mysql", several comments:

* If we do this, shouldn't we check then when validating that when highAvailibility is defined to true the mysql field must be null?

* I still have one concern related to my previous comment https://github.com/3scale/3scale-operator/pull/275#issuecomment-554454571. If we are going to remove the explicit setting of any default for the database field in the system section I'd say we should do the same for the fileStorage section. It just seems there's something wrong when we explicitly set defaults for all other "union types" we have but not for this one.

* What happens with existing customers deployments and their fileStorage(if we decide to change that too) and database sections where the defaults are ALREADY set? From the point of view of this changes existing clients might already have an inconsistent view of the CR. Should we have migration code to "unset" them if they are set to defaults? If we decide to not migrate/modify anything for existing deployments could this lead us to some problem in the future?

cc @eguzki to see this feedback

Reply:
> If we do this, shouldn't we check then when validating that when highAvailibility is defined to true the mysql field must be null?

We could do that. Should we? Is there any use case where the customer wants the following scenario?
* redis databases managed externally
* system relational databases be managed internally

If that would be the case, I would express the desired state like this:

```
spec:                                               
  wildcardDomain: apps.example.com                
  highAvailability:                                 
    enabled: true           
  system:
    database:
      postgresql: {}
```
This desired state is not supported currently. That is why that validation you mention should be done. But if the use case above would be valid, the validation you request would not make sense. 

I will implement that validation. 

> I still have one concern related to my previous comment #275 (comment). If we are going to remove the explicit setting of any default for the database field in the system section I'd say we should do the same for the fileStorage section. It just seems there's something wrong when we explicitly set defaults for all other "union types" we have but not for this one.

I think it is a good idea. I will implement it.

> What happens with existing customers deployments and their fileStorage(if we decide to change that too) and database sections where the defaults are ALREADY set? From the point of view of this changes existing clients might already have an inconsistent view of the CR. Should we have migration code to "unset" them if they are set to defaults? If we decide to not migrate/modify anything for existing deployments could this lead us to some problem in the future?

For customers on previous 3scale releases and `.spec.highAvailability.enabled` set to `true`, the CR looks like this:

```
spec:
  ...
  highAvailability:
    enabled: true
  system:
    database:
      mysql: {}
```
I think the CR should be "cleaned" in the upgrade procedure. I will implement it.